src/code.cloudfoundry.org/loggregator/plumbing/envelope_averager.go

```diff
@@ -0,0 +1,81 @@
+package plumbing
+
+import (
+	"sync/atomic"
+	"time"
+)
+
+// EnvelopeAverager implements google.golang.org/grpc/stats.Handler and keeps
+// track of the average envelope size emitted by a grpc client. It should
+// be constructed with NewEnvelopeAverager.
+//
+// Note: Track is on a very "hot" path.
+// Care should be taken while altering any algorithms, and performance
+// must be considered.
+type EnvelopeAverager struct {
+	// storage must be accessed via atomics
+	storage countAndTotal
+}
+
+// NewEnvelopeAverager creates a new EnvelopeAverager.
+func NewEnvelopeAverager() *EnvelopeAverager {
+	return &EnvelopeAverager{}
+}
+
+// Track takes the given envelope size (in bytes) to use in the current
+// average calculation. It can be called by several go-routines.
+func (a *EnvelopeAverager) Track(envelopeSize int) {
+	for {
+		current := atomic.LoadUint64((*uint64)(&a.storage))
+		newValue := countAndTotal(current).encodeDelta(1, uint32(envelopeSize))
+
+		if atomic.CompareAndSwapUint64((*uint64)(&a.storage), current, uint64(newValue)) {
+			return
+		}
+	}
+}
+
+// Start invokes the given callback at the set interval.
+func (a *EnvelopeAverager) Start(interval time.Duration, f func(average float64)) {
+	go func() {
+		var prevCount, prevTotal uint32
+		for range time.Tick(interval) {
+			current := atomic.LoadUint64((*uint64)(&a.storage))
```

Comment:
I would prefer only one path be able to write. This also prevents `Start()` from having a side effect. If both paths write, then both would have to do a CAS.